File: library-core/src/main/java/no/wtw/android/restserviceutils/resource/Link.java
package no.wtw.android.restserviceutils.resource;

import com.google.gson.annotations.SerializedName;

import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RestTemplate;

import java.io.Serializable;
import java.net.URI;
import java.util.HashMap;
import java.util.Map;

import no.wtw.android.restserviceutils.exceptions.LinkNotResolvedException;
import no.wtw.android.restserviceutils.exceptions.RestServiceException;

public class Link<T> implements Serializable {

    private static final String TAG = Link.class.getSimpleName();

    public static final String LINK_SELF = "self";

    private transient T resource;
    private transient Class<T> clazz;

    @SerializedName("key")
    protected String relation;

    @SerializedName("href")
    protected String url;

    private Map<String, String> queryParams;

    public Link() {
    }

    public Link(String relation, String url) {
        this.relation = relation;
        this.url = url;
    }

    public String getRelation() {
        return relation;
    }

    public String getUrl() {
        return url;
    }

    public T httpGet(final RestTemplate restTemplate) throws RestServiceException {
        if (clazz == null)
            throw new RuntimeException("Class of return object must be set");
        if (getQueryParams() != null)
            return httpGet(restTemplate, getQueryParams());
        return executeHttpCall(() -> {
            resource = restTemplate.getForObject(getUrl(), clazz);
            return resource;
        });
    }

    public T httpGet(final RestTemplate restTemplate, final JsonEncodedQuery query) throws RestServiceException {
        if (clazz == null)
            throw new RuntimeException("Class of return object must be set");
        return executeHttpCall(() -> {
            String url = getUrl().replace("?data=Base64", ""); // TODO: remove this hack
            resource = restTemplate.getForObject(url + "?data=" + query.encode(true), clazz);
            return resource;
        });
    }

    public T httpGet(final RestTemplate restTemplate, final Map<String, String> queryParams) throws RestServiceException {
        if (clazz == null)
            throw new RuntimeException("Class of return object must be set");
        return executeHttpCall(() -> {
            String queryString = "";
            for (String key : queryParams.keySet())
                queryString += key + "=" + queryParams.get(key) + "&";
            resource = restTemplate.getForObject(getUrl() + "?" + queryString.substring(0, queryString.length() - 1), clazz);
            return resource;
        });
    }

    public T httpPut(final RestTemplate restTemplate, final T body) throws RestServiceException {
        return executeHttpCall(() -> {
            HttpEntity<T> requestEntity = new HttpEntity<>(body, new HttpHeaders());
            ResponseEntity<T> responseEntity = restTemplate.exchange(URI.create(getUrl()), HttpMethod.PUT, requestEntity, clazz);
            if (HttpStatus.OK.equals(responseEntity.getStatusCode()))
                resource = responseEntity.getBody();
            return resource;
        });
    }

    public T httpPost(final RestTemplate restTemplate, final Object body) throws RestServiceException {
        return executeHttpCall(() -> restTemplate.postForEntity(getUrl(), body, clazz, new HashMap<>()).getBody());
    }

    public void httpDelete(final RestTemplate restTemplate) throws RestServiceException {
        executeHttpCall((HttpCall<Void>) () -> {
            restTemplate.delete(getUrl());
            return null;
        });
    }

    public T getResource() throws LinkNotResolvedException {
        if (resource == null)
            throw new LinkNotResolvedException();
        return resource;
    }

    public boolean isResolved() {
        return resource != null;
    }

    public void setQueryParams(Map<String, String> queryParams) {
        this.queryParams = queryParams;
    }

    public Map<String, String> getQueryParams() {
        return queryParams;
    }

    public void setClass(Class<T> clazz) {
        this.clazz = clazz;
    }

    private <RT, C extends HttpCall<RT>> RT executeHttpCall(C call) throws RestServiceException {
        try {
            return call.httpCall();
        } catch (Exception e) {
            if (e.getMessage() != null)
                System.out.println(e.getMessage());
            throw RestServiceException.getInstance(e);
        }
    }

    interface HttpCall<RT> {
        RT httpCall() throws Exception;
    }

}

File: library-core/src/main/java/no/wtw/android/restserviceutils/resource/Link.kt
package no.wtw.android.restserviceutils.resource

import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import no.wtw.android.restserviceutils.exceptions.LinkNotResolvedException
import no.wtw.android.restserviceutils.exceptions.RestServiceException
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.Serializable

class Link<T> : Serializable {

    companion object {
        const val LINK_SELF = "self"
    }

    @Transient
    private var _resource: T? = null
    val resource: T
        @Throws(LinkNotResolvedException::class)
        get() = _resource ?: throw LinkNotResolvedException()

    @Transient
    private var clazz: Class<T>? = null

    @SerializedName("key")
    var relation: String? = null
        protected set

    @SerializedName("href")
    var url: String? = null
        protected set

    var queryParams: Map<String, String>? = null

    constructor() {}
    constructor(relation: String?, url: String?) {
        this.relation = relation
        this.url = url
    }

    val isResolved: Boolean
        get() = _resource != null

    fun setClass(clazz: Class<T>?) {
        this.clazz = clazz
    }

    fun httpGet(client: OkHttpClient, gson: Gson): T {
        val qp = queryParams
        if (qp != null)
            return httpGet(client, gson, qp)
        return executeHttpCall {
            if (clazz == null)
                throw RuntimeException("Class of return object must be set")
            val request = Request.Builder().url(url!!).get().build()
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful)
                    throw RestServiceException.from(response)
                val body = response.body?.string() ?: ""
                _resource = gson.fromJson(body, clazz)
                _resource ?: throw NullPointerException("Body is null")
            }
        }
    }

    fun httpGet(client: OkHttpClient, gson: Gson, queryParams: Map<String, String>): T {
        return executeHttpCall {
            if (clazz == null)
                throw RuntimeException("Class of return object must be set")
            var queryString = ""
            for (key in queryParams.keys)
                queryString += key + "=" + queryParams[key] + "&"
            val request = Request.Builder().url(url!! + "?" + queryString.substring(0, queryString.length - 1)).get().build()
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful)
                    throw RestServiceException.from(response)
                val body = response.body?.string() ?: ""
                _resource = gson.fromJson(body, clazz)
                _resource ?: throw NullPointerException("Body is null")
            }
        }
    }

    fun httpGet(client: OkHttpClient, gson: Gson, query: JsonEncodedQuery): T {
        return executeHttpCall {
            if (clazz == null)
                throw RuntimeException("Class of return object must be set")
            val url = url!!.replace("?data=Base64", "") // TODO: remove this hack
            val request = Request.Builder().url(url + "?data=" + query.encode(true)).get().build()
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful)
                    throw RestServiceException.from(response)
                val body = response.body?.string() ?: ""
                _resource = gson.fromJson(body, clazz)
                _resource ?: throw NullPointerException("Body is null")
            }
        }
    }

    fun httpGet(client: OkHttpClient, gson: Gson, urlAppendix: String): T {
        return executeHttpCall {
            if (clazz == null)
                throw RuntimeException("Class of return object must be set")
            val request = Request.Builder().url(url!! + urlAppendix).get().build()
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful)
                    throw RestServiceException.from(response)
                val body = response.body?.string() ?: ""
                _resource = gson.fromJson(body, clazz)
                _resource ?: throw NullPointerException("Body is null")
            }
        }
    }

    fun httpPut(client: OkHttpClient, gson: Gson, data: T): T {
        return executeHttpCall {
            if (clazz == null)
                throw RuntimeException("Class of return object must be set")
            val jsonData = data?.let { gson.toJson(it) } ?: ""
            val mediaType = "application/json; charset=utf-8".toMediaType()
            val requestBody: RequestBody = jsonData.toRequestBody(mediaType)
            val request = Request.Builder().url(url!!).put(requestBody).build()
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful)
                    throw RestServiceException.from(response)
                val body = response.body?.string() ?: ""
                _resource = gson.fromJson(body, clazz)
                _resource ?: throw NullPointerException("Body is null")
            }
        }
    }

    fun httpPost(client: OkHttpClient, gson: Gson, data: Any?): T {
        return executeHttpCall {
            val jsonData = data?.let { gson.toJson(it) } ?: ""
            val mediaType = "application/json; charset=utf-8".toMediaType()
            val requestBody: RequestBody = jsonData.toRequestBody(mediaType)
            val request = Request.Builder().url(url!!).post(requestBody).build()
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful)
                    throw RestServiceException.from(response)
                if (clazz == null) {
                    Unit as T
                } else {
                    val body = response.body?.string() ?: ""
                    gson.fromJson(body, clazz)
                }
            }
        }
    }

    fun httpDelete(client: OkHttpClient, gson: Gson): Unit =
        executeHttpCall {
            val request = Request.Builder().url(url!!).delete().build()
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful)
                    throw RestServiceException.from(response)
            }
        }

    private fun <RT> executeHttpCall(block: () -> RT): RT {
        return try {
            block()
        } catch (e: Exception) {
            if (e.message != null)
                println(e.message)
            throw RestServiceException.from(e)
        }
    }

}